dataclients/kubernetes/routegroupcrd.go

```diff
@@ -0,0 +1,427 @@
+			continue
+		}
+
+		m[si] = true
+		u = append(u, si)
+	}
+
+	return u
+}
+
+func (rg *routeGroupSpec) uniqueHosts() []string {
+	return uniqueStrings(rg.Hosts)
+}
+
+func (rg *routeGroupSpec) validate() error {
+	// has at least one backend:
+	if len(rg.Backends) == 0 {
+		return errRouteGroupWithoutBackend
+	}
+
+	// backends valid and have unique names:
+	backends := make(map[string]bool)
+	for _, b := range rg.Backends {
+		if backends[b.Name] {
+			return backendsWithDuplicateName(b.Name)
+		}
+
+		backends[b.Name] = true
+		if err := b.validate(); err != nil {
+			return invalidBackend(b.Name, err)
+		}
+	}
+
+	hasDefault := len(rg.DefaultBackends) > 0
+	for _, br := range rg.DefaultBackends {
+		if err := br.validate(backends); err != nil {
+			return err
+		}
+	}
+
+	if !hasDefault && len(rg.Routes) == 0 {
+		return errMissingBackendReference
+	}
+
+	for i, r := range rg.Routes {
+		if err := r.validate(hasDefault, backends); err != nil {
+			return invalidRoute(i, err)
+		}
+	}
+
+	return nil
+}
+
+// adding Kubernetes specific backend types here. To be discussed.
+// The main reason is to differentiate between service and external, in a way
+// where we can also use the current global option to decide whether the service
+// should then be converted to LB. Or shall we expect the route group already
+// contain the pod endpoints, and ignore the global option for skipper?
+// --> As CRD we have to lookup endpoints ourselves, maybe via kube.go
+const (
+	serviceBackend = eskip.LBBackend + 1 + iota
+)
+
+func backendTypeFromString(s string) (eskip.BackendType, error) {
+	switch s {
+	case "", "service":
+		return serviceBackend, nil
+	default:
+		return eskip.BackendTypeFromString(s)
+	}
+}
+
+func backendTypeToString(t eskip.BackendType) string {
+	switch t {
+	case serviceBackend:
+		return "service"
+	default:
+		return t.String()
+	}
+}
+
+// UnmarshalJSON creates a new skipperBackend, safe to be called on nil pointer
+func (sb *skipperBackend) UnmarshalJSON(value []byte) error {
+	var p skipperBackendParser
+	if err := json.Unmarshal(value, &p); err != nil {
+		return err
+	}
+
+	var perr error
+	bt, err := backendTypeFromString(p.Type)
+	if err != nil {
+		// we cannot return an error here, because then the parsing of
+		// all route groups would fail. We'll report the error in the
+		// validation phase, only for the containing route group
+		perr = err
+	}
+
+	a, err := loadbalancer.AlgorithmFromString(p.Algorithm)
+	if err != nil {
+		// we cannot return an error here, because then the parsing of
+		// all route groups would fail. We'll report the error in the
+		// validation phase, only for the containing route group
+		perr = err
+	}
+
+	if a == loadbalancer.None {
+		a = loadbalancer.RoundRobin
+	}
+
+	var b skipperBackend
+	b.Name = p.Name
+	b.Type = bt
+	b.Address = p.Address
+	b.ServiceName = p.ServiceName
+	b.ServicePort = p.ServicePort
+	b.Algorithm = a
+	b.Endpoints = p.Endpoints
+	b.parseError = perr
+
+	*sb = b
```

Comment:
Is this not a nil ptr deref similar to https://play.golang.org/p/dT7-sRilS1y

Reply:
https://godoc.org/encoding/json#Unmarshal

the doc can be confusing, because the input arg cannot be nil, but the fields can be nil. I can add here some guards, but there's on reasonable way to test it :)

Reply:
added a guard